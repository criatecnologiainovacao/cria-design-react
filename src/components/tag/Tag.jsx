/* @flow */

import React from 'react';
import { Component, Transition, View } from '../../../libs';
import PropTypes from 'prop-types';

export default class Tag extends Component {
    constructor(props) {
        super(props);

        this.state = {
            visible: true
        }
    }

    handleClose() {
        this.setState({ visible: false }, () => {
            if (this.props.onClose) this.props.onClose();
        });
    }

    handleClick() {
        if (this.props.onClick) this.props.onClick();
    }

    render() {
        const {
            type,
            icon,
            appendIcon,
            hit,
            solid,
            closable,
            round,
            disableTransitions,
            size,
            children
        } = this.props;

        return (
            <Transition name={disableTransitions ? '' : 'cd-zoom-in-center'}>
                <View key={this.state.visible} show={this.state.visible}>
          <span
              style={this.style()}
              onClick={this.handleClick.bind(this)}
              className={
                  this.className(
                      'cd-tag',
                      type && `cd-tag--${type}`,
                      size && `cd-tag--${size}`,
                      {
                          'is-hit': hit,
                          'is-round': round,
                          'is-solid': solid
                      }
                  )
              }
          >
              {icon && <i className={`cd-icon ${icon}`}/>}
              {children && <span>{children}</span>}
              {appendIcon && !closable && <i className={`cd-append-icon ${appendIcon}`}/>}
              {
                  closable &&
                  <i className="cd-tag__close cd-icon-close" onClick={this.handleClose.bind(this)}/>
              }
          </span>
                </View>
            </Transition>
        )
    }

}

Tag.propTypes = {
    type: PropTypes.string,
    icon: PropTypes.string,
    appendIcon: PropTypes.string,
    hit: PropTypes.bool,
    solid: PropTypes.bool,
    closable: PropTypes.bool,
    round: PropTypes.bool,
    disableTransitions: PropTypes.bool,
    size: PropTypes.string,
    onClose: PropTypes.func,
    onClick: PropTypes.func
};