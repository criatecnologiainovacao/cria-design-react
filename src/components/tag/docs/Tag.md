## Dialog

Informs users while preserving the current page state.

### Basic usage

Dialog pops up a dialog box, and it's quite customizable.

:::demo Set the `visible` attribute with a `Boolean`, and Dialog shows when it is `true`. The Dialog has two parts: `Dialog.Body` and `Dialog.Footer`. The optional `title` attribute (empty by default) is for defining a title.

```js
constructor(props) {
  super(props);

  this.state = {
    dialogVisible: false
  };
}

render() {
  return (
    <div>
    </div>
  )
}
```
:::

### Customizations

The content of Dialog can be anything, even a table or a form. This example shows how to use Element Table and Form with Dialog。

:::demo

```js
constructor(props) {
  super(props);

  this.state = {
    dialogVisible2: false,
    dialogVisible3: false,
    form: {
      name: '',
      region: ''
    }
  };

  this.table = {
    columns: [
      {
        label: "Date",
        prop: "date",
        width: 150
      },
      {
        label: "Name",
        prop: "name",
        width: 100
      },
      {
        label: "Address",
        prop: "address"
      }
    ],
    data: [{
      date: '2016-05-02',
      name: 'John Smith',
      address: 'No.1518,  Jinshajiang Road, Putuo District'
    }, {
      date: '2016-05-04',
      name: 'John Smith',
      address: 'No.1518,  Jinshajiang Road, Putuo District'
    }, {
      date: '2016-05-01',
      name: 'John Smith',
      address: 'No.1518,  Jinshajiang Road, Putuo District'
    }, {
      date: '2016-05-03',
      name: 'John Smith',
      address: 'No.1518,  Jinshajiang Road, Putuo District'
    }]
  };
}

render() {
  return (
    <div>
     
    </div>
  )
}
```
:::

### Attributes

| Attribute      | Description          | Type      | Accepted Values       | Default  |
|---------- |-------------- |---------- |--------------------------------  |-------- |
| title     | title of Dialog. Can also be passed with a named slot (see the following table) | string    | — | — |
| size      | size of Dialog | string    | tiny/small/large/full | small |
| top      | value for `top` of Dialog CSS, works when `size` is not `full` | string    | — | 15% |
| modal     | whether a mask is displayed | boolean   | — | true |
| modalAppendToBody     | whether to append modal to body element. If false, the modal will be appended to Dialog's parent element | boolean   | — | true |
| lockScroll     | whether scroll of body is disabled while Dialog is displayed | boolean   | — | true |
| customClass      | custom class names for Dialog | string    | — | — |
| closeOnClickModal | whether the Dialog can be closed by clicking the mask | boolean    | — | true |
| closeOnPressEscape | whether the Dialog can be closed by pressing ESC | boolean    | — | true |
| showClose | whether to show a close button | boolean    | — | true |

### Events
| Event Name | Description | Parameters |
|---------- |-------- |---------- |
| onOpen | triggers when the Dialog opens | — |
| onClose | triggers when the Dialog closes | — |
