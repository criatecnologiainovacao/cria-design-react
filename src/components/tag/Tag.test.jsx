import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import renderer from 'react-test-renderer';

import Tag from './Tag';

describe('Tag tests', () => {

    it('should create default tag', () => {

        const tag = shallow(<Tag>TEST</Tag>);

        expect(tag.find('.cd-tag').text()).toEqual('TEST');
    });

    it('should default tag render correctly', () => {

        const tag = renderer
            .create(<Tag>TEST</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should typed tags have correctly classes', () => {

        const successTag = shallow(<Tag type="success">TEST</Tag>);
        const infoTag = shallow(<Tag type="info">TEST</Tag>);
        const warningTag = shallow(<Tag type="warning">TEST</Tag>);
        const dangerTag = shallow(<Tag type="danger">TEST</Tag>);

        expect(successTag.find('.cd-tag')).toHaveClassName('cd-tag--success');
        expect(infoTag.find('.cd-tag')).toHaveClassName('cd-tag--info');
        expect(warningTag.find('.cd-tag')).toHaveClassName('cd-tag--warning');
        expect(dangerTag.find('.cd-tag')).toHaveClassName('cd-tag--danger');
    });

    it('should typed tags render correctly', () => {

        const successTag = renderer.create(<Tag type="success">TEST</Tag>).toJSON();
        const infoTag = renderer.create(<Tag type="info">TEST</Tag>).toJSON();
        const warningTag = renderer.create(<Tag type="warning">TEST</Tag>).toJSON();
        const dangerTag = renderer.create(<Tag type="danger">TEST</Tag>).toJSON();

        expect(successTag).toMatchSnapshot();
        expect(infoTag).toMatchSnapshot();
        expect(warningTag).toMatchSnapshot();
        expect(dangerTag).toMatchSnapshot();
    });

    it('should tag with icon be properly created', () => {

        const tag = shallow(<Tag icon="cd-icon-search">TESTE</Tag>);

        expect(tag.find('i.cd-icon').exists()).toBe(true);
        expect(tag.find('i.cd-icon-search').exists()).toBe(true);
    });

    it('should tag with icon render correctly', () => {

        const tag = renderer
            .create(<Tag icon="cd-icon-search">TESTE</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should tag with appendIcon be properly created', () => {

        const tag = shallow(<Tag appendIcon="cd-icon-search">TESTE</Tag>);

        expect(tag.find('i.cd-append-icon').exists()).toBe(true);
        expect(tag.find('i.cd-icon-search').exists()).toBe(true);
    });

    it('should tag with appendIcon render correctly', () => {

        const tag = renderer
            .create(<Tag appendIcon="cd-icon-search">TESTE</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should tag with appendIcon render correctly (without label)', () => {

        const tag = renderer
            .create(<Tag appendIcon="cd-icon-search"/>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should closable tag have close icon', () => {
        const tag = shallow(<Tag closable>TEST</Tag>);

        expect(tag.find('i.cd-tag__close').exists()).toBe(true);
    });

    it('should closable tag render correctly', () => {
        const tag = renderer
            .create(<Tag closable>TEST</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should have hit class', () => {
        const tag = shallow(<Tag hit>TEST</Tag>);

        expect(tag.find('.cd-tag').first().hasClass('is-hit')).toBeTruthy();
    });

    it('should hit tag render correctly', () => {
        const tag = renderer
            .create(<Tag hit>TEST</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should heve solid class', () => {
        const tag = shallow(<Tag solid>TEST</Tag>);

        expect(tag.find('.cd-tag').first().hasClass('is-solid')).toBeTruthy();
    });

    it('should solid tag render correctly', () => {
        const tag = renderer
            .create(<Tag solid>TEST</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should round tag have properly class', () => {
        const tag = shallow(<Tag round>TEST</Tag>);

        expect(tag.find('.cd-tag').first().hasClass('is-round')).toBeTruthy();
    });

    it('should round tag render correctly', () => {
        const tag = renderer
            .create(<Tag round>TEST</Tag>)
            .toJSON();

        expect(tag).toMatchSnapshot();
    });

    it('should simulate click when onClose is present', () => {
        const onClose = sinon.spy();
        const tag = shallow(React.createElement(
            Tag,
            { type: 'primary', closable: true, onClose: onClose },
            'TEST'
        ));

        tag.find('i.cd-tag__close').simulate('click');

        expect(onClose.calledOnce).toBe(true);
    });
});