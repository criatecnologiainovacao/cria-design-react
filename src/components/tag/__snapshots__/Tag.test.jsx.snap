// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag tests should closable tag render correctly 1`] = `
<span
  className="cd-tag"
  style={Object {}}
>
  <span>
    TEST
  </span>
  <i
    className="cd-tag__close cd-icon-close"
    onClick={[Function]}
  />
</span>
`;

exports[`Tag tests should default tag render correctly 1`] = `
<span
  className="cd-tag"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should hit tag render correctly 1`] = `
<span
  className="cd-tag is-hit"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should round tag render correctly 1`] = `
<span
  className="cd-tag is-round"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should solid tag render correctly 1`] = `
<span
  className="cd-tag is-solid"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should tag with appendIcon render correctly (without label) 1`] = `
<span
  className="cd-tag"
  style={Object {}}
>
  <i
    className="cd-append-icon cd-icon-search"
  />
</span>
`;

exports[`Tag tests should tag with appendIcon render correctly 1`] = `
<span
  className="cd-tag"
  style={Object {}}
>
  <span>
    TESTE
  </span>
  <i
    className="cd-append-icon cd-icon-search"
  />
</span>
`;

exports[`Tag tests should tag with icon render correctly 1`] = `
<span
  className="cd-tag"
  style={Object {}}
>
  <i
    className="cd-icon cd-icon-search"
  />
  <span>
    TESTE
  </span>
</span>
`;

exports[`Tag tests should typed tags render correctly 1`] = `
<span
  className="cd-tag cd-tag--success"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should typed tags render correctly 2`] = `
<span
  className="cd-tag cd-tag--info"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should typed tags render correctly 3`] = `
<span
  className="cd-tag cd-tag--warning"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;

exports[`Tag tests should typed tags render correctly 4`] = `
<span
  className="cd-tag cd-tag--danger"
  style={Object {}}
>
  <span>
    TEST
  </span>
</span>
`;
